miner/src/service_transaction_checker.rs

```diff
@@ -16,37 +16,111 @@
-	pub fn check<C: CallContract + RegistryInfo>(&self, client: &C, tx: &SignedTransaction) -> Result<bool, String> {
+	/// Checks if given address in tx is whitelisted to send service transactions.
+	pub fn check<C: CallContract + RegistryInfo + ChainInfo>(&self, client: &C, tx: &SignedTransaction) -> Result<bool, String> {
 		let sender = tx.sender();
-		let hash = tx.hash();
-
 		// Skip checking the contract if the transaction does not have zero gas price
 		if !tx.gas_price.is_zero() {
 			return Ok(false)
 		}
 
-		let address = client.registry_address(SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME.to_owned(), BlockId::Latest)
+		self.check_address(client, sender)
+	}
+
+	/// Checks if given address is whitelisted to send service transactions.
+	pub fn check_address<C: CallContract + RegistryInfo + ChainInfo>(&self, client: &C, sender: Address) -> Result<bool, String> {
+		let cache = self.certified_addresses_cache.cache.try_read();
+		if (client.chain_info().best_block_number as u64 == self.certified_addresses_cache.last_block) && cache.is_some() {
+			if let Some(allowed) = cache.unwrap().get(&sender) {
+				return Ok(*allowed);
+			}
+		}
+		let contract_address = client.registry_address(SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME.to_owned(), BlockId::Latest)
 			.ok_or_else(|| "contract is not configured")?;
+		trace!(target: "txqueue", "Checking service transaction checker contract from {}", sender);
+		let (data, decoder) = service_transaction::functions::certified::call(sender);
+		let value = client.call_contract(BlockId::Latest, contract_address, data)?;
+		decoder.decode(&value).and_then(|allowed| {
+			let cache = self.certified_addresses_cache.cache.try_write();
+			if cache.is_some() {
+				cache.unwrap().insert(sender, allowed);
+			};
+			Ok(allowed)
+		}).map_err(|e| e.to_string())
+	}
 
-		trace!(target: "txqueue", "[{:?}] Checking service transaction checker contract from {}", hash, sender);
+	/// Refresh certified addresses cache
+	pub fn refresh_cache<C: CallContract + RegistryInfo + ChainInfo>(&mut self, client: &C) -> Result<bool, String> {
+		trace!(target: "txqueue", "Refreshing certified addresses cache");
+		let contract_address = client.registry_address(SERVICE_TRANSACTION_CONTRACT_REGISTRY_NAME.to_owned(), BlockId::Latest);
+		if contract_address.is_none() {
+			return Ok(false)
+		}
+		let mut updated_addresses: BTreeMap<Address, bool> = BTreeMap::default();
+		let cache = self.certified_addresses_cache.cache.try_read();
+		if cache.is_some() {
+			for (address, allowed_before) in cache.unwrap().iter() {
+				let (data, decoder) = service_transaction::functions::certified::call(*address);
+				let value = client.call_contract(BlockId::Latest, contract_address.unwrap(), data)?;
+				let allowed = decoder.decode(&value).map_err(|e| e.to_string())?;
+				if *allowed_before != allowed {
+					updated_addresses.insert(*address, allowed);
+				}
+			};
+			let cache = self.certified_addresses_cache.cache.try_write();
```

Comment:
I would suggest a little bit different logic to refresh the cache:
```rust
// replace the cache with an empty list,
// since it's not recent it won't be used anyway.
let mut cache = mem::replace(&mut self.cache.write(), HashMap::default());
let addresses: Vec<_> = cache.keys().collect();
for address in addresses {
   // make a call
   let is_allowed = ...;
   // update the entry
   cache.insert(address, is_allowed);
}
mem::replace(&mut self.cache.write(),  cache);
```
This means that during the update cache is not used at all.

Reply:
On a second though if the cache is reused between client and miner we would get into a potential attack vector if the cache is never cleared and not limited.